File: src/main/java/amerifrance/guideapi/objects/Book.scala
package amerifrance.guideapi.objects

import java.awt.Color
import java.util

import amerifrance.guideapi.objects.abstraction.AbstractCategory
import net.minecraft.util.{ResourceLocation, StatCollector}

class Book(categoryList: util.List[AbstractCategory] = new util.ArrayList[AbstractCategory](), unlocTitle: String, bookColor: Color = new Color(171, 80, 30), unlocWelcomeMessage: String, pageTextureLoc: ResourceLocation, outlineTextureLoc: ResourceLocation) {

  var categories: util.List[AbstractCategory] = categoryList
  var unlocalizedTitle: String = unlocTitle
  var color: Color = bookColor
  var welcomeMessage: String = unlocWelcomeMessage
  var pageTexture: ResourceLocation = pageTextureLoc
  var outlineTexture: ResourceLocation = outlineTextureLoc

  def addCategory(category: AbstractCategory) = {
    this.categories.add(category)
  }

  def addCategoryList(list: util.List[AbstractCategory]) = {
    this.categories.addAll(list)
  }

  def removeCategory(category: AbstractCategory) = {
    this.categories.remove(category)
  }

  def removeCategoryList(list: util.List[AbstractCategory]) = {
    this.categories.removeAll(list)
  }

  def getLocalizedName(): String = {
    return StatCollector.translateToLocal(unlocalizedTitle)
  }

  def getLocalizedWelcomeMessage(): String = {
    return StatCollector.translateToLocal(welcomeMessage)
  }

  def canEqual(other: Any): Boolean = other.isInstanceOf[Book]

  override def equals(other: Any): Boolean = other match {
    case that: Book =>
      (that canEqual this) &&
        categories == that.categories &&
        unlocalizedTitle == that.unlocalizedTitle &&
        color == that.color &&
        welcomeMessage == that.welcomeMessage &&
        pageTexture == that.pageTexture &&
        outlineTexture == that.outlineTexture
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(categories, unlocalizedTitle, color, welcomeMessage, pageTexture, outlineTexture)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}

File: src/main/java/amerifrance/guideapi/objects/Book.java
package amerifrance.guideapi.objects;

import amerifrance.guideapi.objects.abstraction.CategoryAbstract;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.StatCollector;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class Book {

    public List<CategoryAbstract> categoryList = new ArrayList<CategoryAbstract>();
    public String unlocBookTitle;
    public String unlocWelcomeMessage;
    public ResourceLocation pageTexture;
    public ResourceLocation outlineTexture;
    public Color bookColor = new Color(171, 70, 30);

    /**
     * @param categoryList        - Category List with all your information
     * @param unlocBookTitle      - Unlocalized name for a book title
     * @param unlocWelcomeMessage - Unlocalized welcome message
     * @param pageTexture         - Texture for book's page
     * @param outlineTexture      - Texture for book outline
     * @param bookColor           - Color of book
     */
    public Book(List<CategoryAbstract> categoryList, String unlocBookTitle, String unlocWelcomeMessage, ResourceLocation pageTexture, ResourceLocation outlineTexture, Color bookColor) {
        this.categoryList = categoryList;
        this.unlocBookTitle = unlocBookTitle;
        this.unlocWelcomeMessage = unlocWelcomeMessage;
        this.pageTexture = pageTexture;
        this.outlineTexture = outlineTexture;
        this.bookColor = bookColor;
    }

    /**
     * @param category - Add this category
     */
    public void addCategory(CategoryAbstract category) {
        this.categoryList.add(category);
    }

    /**
     * @param category - Remove this category
     */
    public void removeCategory(CategoryAbstract category) {
        this.categoryList.remove(category);
    }

    /**
     * @param categories - Add these categories
     */
    public void addCategoryList(List<CategoryAbstract> categories) {
        this.categoryList.addAll(categories);
    }

    /**
     * @param categories - Remove these categories
     */
    public void removeCategoryList(List<CategoryAbstract> categories) {
        this.categoryList.removeAll(categories);
    }

    /**
     * @return - Localized book title
     */
    public String getLocalizedBookTitle() {
        return StatCollector.translateToLocal(unlocBookTitle);
    }

    /**
     * @return - Localized welcome message
     */
    public String getLocalizedWelcomeMessage() {
        return StatCollector.translateToLocal(unlocWelcomeMessage);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Book book = (Book) o;
        if (bookColor != null ? !bookColor.equals(book.bookColor) : book.bookColor != null) return false;
        if (categoryList != null ? !categoryList.equals(book.categoryList) : book.categoryList != null) return false;
        if (outlineTexture != null ? !outlineTexture.equals(book.outlineTexture) : book.outlineTexture != null)
            return false;
        if (pageTexture != null ? !pageTexture.equals(book.pageTexture) : book.pageTexture != null) return false;
        if (unlocBookTitle != null ? !unlocBookTitle.equals(book.unlocBookTitle) : book.unlocBookTitle != null)
            return false;
        if (unlocWelcomeMessage != null ? !unlocWelcomeMessage.equals(book.unlocWelcomeMessage) : book.unlocWelcomeMessage != null)
            return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = categoryList != null ? categoryList.hashCode() : 0;
        result = 31 * result + (unlocBookTitle != null ? unlocBookTitle.hashCode() : 0);
        result = 31 * result + (unlocWelcomeMessage != null ? unlocWelcomeMessage.hashCode() : 0);
        result = 31 * result + (pageTexture != null ? pageTexture.hashCode() : 0);
        result = 31 * result + (outlineTexture != null ? outlineTexture.hashCode() : 0);
        result = 31 * result + (bookColor != null ? bookColor.hashCode() : 0);
        return result;
    }
}
